server/src/main/java/org/elasticsearch/gateway/PersistedClusterStateService.java

```diff
@@ -291,17 +295,46 @@ public static void overrideVersion(Version newVersion, Path... dataPaths) throws
      * Loads the available on-disk cluster state. Returns {@link OnDiskState#NO_ON_DISK_STATE} if no such state was found.
      */
     public OnDiskState loadOnDiskState() throws IOException {
+        return loadOnDiskState(true);
+    }
+
+    /**
+     * Loads the available on-disk cluster state. Returns {@link OnDiskState#NO_ON_DISK_STATE} if no such state was found.
+     * @param checkClean whether to check the index for corruption before loading, only for tests
+     */
+    OnDiskState loadOnDiskState(boolean checkClean) throws IOException {
         OnDiskState onDiskState = OnDiskState.NO_ON_DISK_STATE;
 
         final Path indexPath = dataPath.resolve(METADATA_DIRECTORY_NAME);
         if (Files.exists(indexPath)) {
-            try (Directory directory = createDirectory(indexPath);
-                 DirectoryReader directoryReader = DirectoryReader.open(directory)) {
-                onDiskState = loadOnDiskState(dataPath, directoryReader);
+            try (Directory directory = createDirectory(indexPath)) {
+                if (checkClean) {
+                    try (BytesStreamOutput outputStream = new BytesStreamOutput()) {
+                        final boolean isClean;
+                        try (PrintStream printStream = new PrintStream(outputStream, true, StandardCharsets.UTF_8);
+                             CheckIndex checkIndex = new CheckIndex(directory)) {
+                            checkIndex.setInfoStream(printStream);
+                            checkIndex.setChecksumsOnly(true);
+                            isClean = checkIndex.checkIndex().clean;
+                        }
 
-                if (nodeId.equals(onDiskState.nodeId) == false) {
-                    throw new IllegalStateException("unexpected node ID in metadata, found [" + onDiskState.nodeId +
-                        "] in [" + dataPath + "] but expected [" + nodeId + "]");
+                        if (isClean == false) {
+                            if (logger.isErrorEnabled()) {
+                                outputStream.bytes().utf8ToString().lines().forEach(l -> logger.error("checkIndex: {}", l));
+                            }
+                            throw new IllegalStateException("metadata index at [" + dataPath +
+                                "] has been changed by an external force after it was last written by Elasticsearch and is now unreadable");
+                        }
+                    }
+                }
+
+                try (DirectoryReader directoryReader = DirectoryReader.open(directory)) {
+                    onDiskState = loadOnDiskState(dataPath, directoryReader);
+
+                    if (nodeId.equals(onDiskState.nodeId) == false) {
+                        throw new IllegalStateException("unexpected node ID in metadata, found [" + onDiskState.nodeId +
```

Comment:
Same comment on `metadata` as above, perhaps `global state metadata`?

Reply:
You had to be doing something pretty weird to hit this message anyway, and doubly so today since the node IDs that we're comparing both come from the user-data from the latest commit of this index. Harmonised with the other message in 9e89fed6f2b2628d1ac09af6f631639ee05364ef.